tools/tune/eq/gui.m

```diff
@@ -0,0 +1,476 @@
+  eq = eq_defaults();
+  % UI panels, we have to use absolute positioning
+  %
+  % +-------+-----------+-------+--+--+--+--+--+----------+
+  % | Label | Dropdown  |       |  |  |  |  |  | checkbox |
+  % +-------+-----------+-------+--+--+--+--+--+----------+
+  % | Label | input box | label | slider                  |
+  % +-------+-----------+-------+-------------------------+
+  % | Label | input box | label | slider                  |
+  % +-------+-----------+-------+-------------------------+
+  % | Label | input box | label | slider                  |
+  % +-------+-----------+-------+-------------------------+
+
+  EQ(index, channel).pane = uipanel ("parent", main_panes(channel).pane, "position", position);
+  p = EQ(index, channel).pane;
+  % octave is silly and doesn't let one position controls relatively inside a uipanel
+  EQ(index, channel).gain.label = uicontrol( ...
+      "style", "text", ...
+      "parent", p, ...
+      "position", [0 10 40 30], ...
+      "visible", "off", ...
+      "string", "Gain");
+  EQ(index, channel).gain.edit = uicontrol( ...
+      "style", "edit", ...
+      "parent", p, ...
+      "position", [40 10 90 30], ...
+      "string", "0", ...
+      "userdata", struct( ...
+          "index", index, ...
+          "channel", channel, ...
+          "prefix", "gain", ...
+          "EQ", EQ, ...
+          "main_panes", main_panes), ...
+      "visible", "off", ...
+      "callback", @edit_callback);
+  EQ(index, channel).gain.units = uicontrol( ...
+      "style", "text", ...
+      "parent", p, ...
+      "string", "dB", ...
+      "visible", "off", ...
+      "position", [140 10 17 30]);
+  EQ(index, channel).gain.slider = uicontrol( ...
+      "style", "slider", ...
+      "parent", p, ...
+      "position", [160 10 200 30], ...
+      "min", -40, ...
+      "max", 40, ...
+      "userdata", struct(
+          "index", index, ...
+          "channel", channel, ...
+          "prefix", "gain", ...
+          "EQ", EQ, ...
+          "main_panes", main_panes), ...
+      "visible", "off", ...
+      "callback", @slider_callback);
+
+  EQ(index, channel).Q.label = uicontrol( ...
+      "style", "text", ...
+      "parent", p, ...
+      "position", [0 50 50 30], ...
+      "visible", "off", ...
+      "string", "Q");
+  EQ(index, channel).Q.edit = uicontrol( ...
+      "style", "edit", ...
+      "parent", p, ...
+      "position", [40 50 90 30], ...
+      "string", "1", ...
+      "userdata", struct( ...
+          "index", index, ...
+          "channel", channel, ...
+          "prefix", "Q", ...
+          "EQ", EQ, ...
+          "main_panes", main_panes), ...
+      "visible", "off", ...
+      "callback", @edit_callback);
+  EQ(index, channel).Q.slider = uicontrol( ...
+      "style", "slider",                 ...
+      "parent", p,                       ...
+      "position", [160 50 200 30],       ...
+      "min", 0,                          ...
+      "max", 1000,                       ...
```

Comment:
Is such high max Q value needed? The pushing up/down the Q value doesn't really work due to the high step size. If the Q max could be e.g. 50 this control would work better. Or make the slider logaritmic to have better precision for the low values.

Reply:
I agree, this should be a log slider. I will test it out again but I remember having trouble trying to get it work due to the GUI elements overflowing. will try again as that is what our web based tool does so it would be best to mirror that.

Reply:
The tool is pretty usable with numerically entered Q value so also this is not critical.

I don't know if it's possible to print/enter to text field a non 1:1 relation. If the slider range would be e.g. 0..log2(1000/0.1), then something like 0.1*2^slider would provide Q value range 0..1000 and first third of the slider would something where typical peak/notch filters are. But not sure if this possible to do without redoing GUI elements.

Reply:
That should be possible, I think I had my math wrong because I think my slider range was exponential which was causing OpenGL to get upset. I will see if I can test it out before I head home for the holidays.

Reply:
Tested it out and it works beautifully thanks for the math. I remember struggling with this quite a bit, clearly it was because I was going the wrong way with the math.